data/resources.go

```diff
@@ -9,104 +9,243 @@ import (
+		p.UnpenalizedHourlyProduction = p.UnpenalizedHourlyProduction.Add(production)
+
+		capacity := b.Capacity()
+		p.Capacity = p.Capacity.Add(capacity)
+	}
+}
+
+func (p *PlanetWithResourceInfo) applyProductionPenalties() {
+	p.EnergyPenalty = p.EnergyProduction < p.EnergyConsumption
+	p.PopulationPenalty = p.PopulationSize < p.PopulationEmployment
+
+	energyProductivity := productivity(p.EnergyProduction, p.EnergyConsumption)
+	populationProductivity := productivity(p.PopulationSize, p.PopulationEmployment)
+	productivity := math.Min(energyProductivity, populationProductivity)
+
+	p.PenalizedHourlyProduction = p.UnpenalizedHourlyProduction.MulFloat64(productivity)
+
+	// TODO: it would be nice that if there is a scarcity of hydrogen, productivity of
+	// 		 hydrogen will be reduced last to save the population.
+}
+
+func (p *PlanetWithResourceInfo) calcNewResourcesAndProduction(now time.Time) {
+	metalFreeSpace := MaxInt64(p.Capacity.Metal-p.Metal, 0)
+	hydroFreeSpace := MaxInt64(p.Capacity.Hydrogen-p.Hydrogen, 0)
+	silicFreeSpace := MaxInt64(p.Capacity.Silica-p.Silica, 0)
+
+	duration := now.Sub(p.LastResourceUpdate)
+	metalProduction := p.PenalizedHourlyProduction.Metal * int64(duration) / int64(time.Hour)
+	hydroProduction := p.PenalizedHourlyProduction.Hydrogen * int64(duration) / int64(time.Hour)
+	silicProduction := p.PenalizedHourlyProduction.Silica * int64(duration) / int64(time.Hour)
+
+	metalProduction = MinInt64(metalProduction, metalFreeSpace)
+	hydroProduction = MinInt64(hydroProduction, hydroFreeSpace)
+	silicProduction = MinInt64(silicProduction, silicFreeSpace)
+
+	p.Metal += metalProduction
+	p.Hydrogen += hydroProduction
+	p.Silica += silicProduction
+	p.LastResourceUpdate = now
+
+	// Recalculate to get the new hourly production (for display in the ui)
+	metalFreeSpace = MaxInt64(p.Capacity.Metal-p.Metal, 0)
+	hydroFreeSpace = MaxInt64(p.Capacity.Hydrogen-p.Hydrogen, 0)
+	silicFreeSpace = MaxInt64(p.Capacity.Silica-p.Silica, 0)
+
+	p.MetalStorageFull = metalFreeSpace == 0
+	p.HydrogenStorageFull = hydroFreeSpace == 0
+	p.SilicaStorageFull = silicFreeSpace == 0
+
+	p.PenalizedHourlyProduction.Metal = MinInt64(p.PenalizedHourlyProduction.Metal, metalFreeSpace)
+	p.PenalizedHourlyProduction.Hydrogen = MinInt64(p.PenalizedHourlyProduction.Hydrogen, hydroFreeSpace)
+	p.PenalizedHourlyProduction.Silica = MinInt64(p.PenalizedHourlyProduction.Silica, silicFreeSpace)
+}
+
+func productivity(available, usage int64) float64 {
+	if usage == 0 {
 		return 1
 	}
-	return float64(prod) / float64(cons)
+	return math.Min(float64(available)/float64(usage), 1)
 }
 
-// GetEnergyProd calculates the current energy production
-func GetEnergyProd(p *ent.Planet) int64 {
-	solarProd := 1500 * int64(p.SolarProdLevel) * int64(math.Pow(1.1, float64(p.SolarProdLevel)))
-	return solarProd
+func solarPlantEnergyProduction(level int) int64 {
+	const initial = 30
+	const base = 1.25
+	l := float64(level + 1) // The player gets one production level for free
```

Comment:
so solar plant will appear as level 1 or it's a hidden free level?

Reply:
Hidden. It will show level 0 but you get some energy from the start. (from your colonization ark or something).